subsys/debug/coredump/coredump_backend_nrf_flash_partition.c

```diff
@@ -0,0 +1,347 @@
+#define FLASH_WRITE_SIZE DT_PROP(FLASH_NODE, write_block_size)
+#if DT_NODE_HAS_PROP(FLASH_NODE, erase_block_size)
+#define FLASH_ERASE_SIZE DT_PROP(FLASH_NODE, erase_block_size)
+#else
+#define FLASH_ERASE_SIZE FLASH_WRITE_SIZE
+#endif
+
+#define PARTITION_OFFSET FIXED_PARTITION_OFFSET(PARTITION_LABEL)
+#define PARTITION_SIZE	 FIXED_PARTITION_SIZE(PARTITION_LABEL)
+#define PARTITION_ADDR	 (FLASH_ADDR + PARTITION_OFFSET)
+
+#if (PARTITION_OFFSET % FLASH_ERASE_SIZE) != 0
+#error "Core dump partition unaligned to erase block size"
+#endif
+
+struct header {
+	uint8_t magic[4];    /* "CD01" */
+	uint32_t offset;     /* Core dump data start */
+	uint32_t size;	     /* Core dump data size */
+	uint16_t dump_crc;   /* Core dump data CRC16 */
+	uint16_t header_crc; /* Header CRC16 (up to this field) */
+} __packed;
+
+static const uint8_t MAGIC[4] = {'C', 'D', '0', '1'};
+
+enum {
+	HEADER_SIZE = ROUND_UP(sizeof(struct header), FLASH_WRITE_SIZE),
+};
+
+static int write_error;			    /* Error occurred when writing a core dump */
+static uint8_t write_buf[FLASH_WRITE_SIZE]; /* Write buffer to assure aligned flash access */
+static size_t write_buf_pos;		    /* # of dump data bytes buffered in the write buffer */
+static size_t write_pos;		    /* # of dump data bytes already written to flash */
+static uint16_t dump_crc;		    /* CRC16 of already written or buffered bytes */
+
+static inline const struct header *get_stored_header(void)
+{
+	return (const struct header *)(PARTITION_ADDR);
+}
+
+static inline const uint8_t *get_stored_dump(const struct header *header)
+{
+	return (const uint8_t *)(PARTITION_ADDR) + header->offset;
+}
+
+static inline uint16_t calc_header_crc(const struct header *header)
+{
+	return crc16_ccitt(0xffff, (const uint8_t *)header, offsetof(struct header, header_crc));
+}
+
+static inline uint16_t calc_dump_crc(const struct header *header)
+{
+	return crc16_ccitt(0xffff, get_stored_dump(header), header->size);
+}
+
+static inline bool validate_header(const struct header *header)
+{
+	return (memcmp(header->magic, MAGIC, sizeof(MAGIC)) == 0) &&
+	       (calc_header_crc(header) == header->header_crc);
+}
+
+static inline bool validate_dump(const struct header *header)
+{
+	return calc_dump_crc(header) == header->dump_crc;
+}
+
+static inline bool is_within_partition(uint32_t offset, size_t size)
+{
+	return (offset < PARTITION_SIZE) && (size <= PARTITION_SIZE - offset);
+}
+
+static void write(uint32_t offset, const uint8_t *data, size_t size)
+{
+	if (write_error != 0) {
+		return;
+	}
+
+	if (!is_within_partition(offset, size)) {
+		write_error = -ENOMEM;
+		return;
+	}
+
+#ifdef CONFIG_NRFX_NVMC
+	for (uint32_t i = 0; i < size; i += sizeof(uint32_t)) {
+		nrfx_nvmc_word_write(PARTITION_OFFSET + offset + i, UNALIGNED_GET(&data[i]));
+	}
+	while (!nrfx_nvmc_write_done_check()) {
+	}
+#elif defined(CONFIG_NRFX_RRAMC)
+	nrf_rramc_config_t config = {.mode_write = true, .write_buff_size = 0};
```

Comment:
Splitting this into multiple lines would increase readability.

Reply:
Done